src/integrators/stepsize_controllers.jl

```diff
@@ -19,27 +19,85 @@ end
 
 
 function stepsize_controller!(integrator::SDEIntegrator, alg::TauLeaping)
-    nothing
+  nothing  # Post-leap adjustment happens in perform_step!
 end
 
 function step_accept_controller!(integrator::SDEIntegrator, alg::TauLeaping)
+  if alg.adaptive
     integrator.q = min(integrator.opts.gamma / integrator.EEst, integrator.opts.qmax)
     return integrator.dt * integrator.q
+  else
+    return integrator.dt
+  end
 end
 
 function step_reject_controller!(integrator::SDEIntegrator, alg::TauLeaping)
+  if alg.adaptive
     integrator.dt = integrator.opts.gamma * integrator.dt / integrator.EEst
+  end
 end
 
-
+# CaoTauLeaping: Pre-leap τ computation
 function stepsize_controller!(integrator::SDEIntegrator, alg::CaoTauLeaping)
-    nothing
+  if !alg.adaptive
+    return
+  end
+  
+  @unpack u, p, t, P, opts, c = integrator
+  cache = integrator.cache
+
+  P === nothing && error("CaoTauLeaping requires a JumpProblem with a RegularJump")
+  
+  # Handle both constant and mutable caches
+  if isa(cache, CaoTauLeapingConstantCache)
+    rate = P.cache.rate(u, p, t)  # Compute propensities directly
+    mu = zero(u)
+    sigma2 = zero(u)
+  else  # CaoTauLeapingCache
+    @unpack mu, sigma2, rate = cache
+    P.cache.rate(rate, u, p, t)  # Compute propensities into cache
+    fill!(mu, zero(eltype(mu)))
+    fill!(sigma2, zero(eltype(sigma2)))
+  end
+
+  # Infer ν_ij using c by applying unit counts for each reaction
+  num_reactions = length(rate)
+  ν = zeros(eltype(u), length(u), num_reactions)
+  unit_counts = zeros(eltype(rate), num_reactions)
+  for j in 1:num_reactions
+    unit_counts[j] = 1
+    c(ν[:, j], u, p, t, unit_counts, nothing)  # ν[:, j] is the change vector for reaction j
+    unit_counts[j] = 0  # Reset
+  end

```

Comment:
This is fine for now, but we really should be storing that differently in the jumps. @isaacsas this is worth thinking about.

Reply:
I think it would be worthwhile to make leaping work with individually defined jumps too. Split-step methods require that anyways.

But that wouldn’t solve this issue for general non-mass action jumps.

Reply:
and we'd need two implementations of each leaping? I'm thinking the interface should maybe require more information in the jump building, under the assumption users will be using something like Catalyst rather than building most jumps by hand. To do some of this optimally we have to shift been a few different representations, where for a user this would be redundant information to the solver it's different views that have different optimal uses.

Reply:
For methods that work on the full vector at once, can’t we can manually build the needed function from the list of jumps? So those methods can stick with the current interface, but then we could support methods that really need jump-by-jump rate/affect access too. 

But yes, Catalyst should build the needed inputs for different representations.

Reply:
> For methods that work on the full vector at once, can’t we can manually build the needed function from the list of jumps?

You get a lot of compile time trying to inline all of the small functions into the large one. Also there's a lack of fusion. So it's quite slow to do it that way.

I think the point is, some times you need jump-by-jump information, and sometimes you want the aggregate information (all tau leap steppers will use an aggregate function a lot of the time), and those are the same information in two very different computational representations, and building one from the other isn't easy to do from the implementation but it is trivial to do symbolically.

Reply:
Definitely, so it should be preferred that methods that work on the full vector get a single input function, but having a fallback would be useful. Catalyst/MTK, should just generate both inputs (or whichever is most appropriate for the selected integrator -- we can add a function to query the preferred input type).